File: sd/slcanx-cpp/examples/02_periodic.cpp
#include "slcanx.hpp"
#include <iostream>
#include <thread>
#include <chrono>

using namespace slcanx;

int main(int argc, char** argv) {
    std::string port = "COM3";
    if (argc > 1) port = argv[1];

    Slcanx slcan(port);
    slcan.close_channel(0);
    slcan.set_bitrate(0, 500000);
    slcan.open_channel(0);

    slcan.set_rx_callback([](uint8_t ch, const CanFrame& frame) {
        std::cout << "Rx Ch" << (int)ch << ": ID=" << std::hex << frame.id << std::dec << std::endl;
    });

    std::cout << "Starting periodic send..." << std::endl;
    
    int cnt = 0;
    while(true) {
        std::vector<uint8_t> data = {0, 1, 2, (uint8_t)cnt};
        CanFrame frame = CanFrame::new_std(0x100, data);
        slcan.send(0, frame);
        
        cnt++;
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return 0;
}

File: sd/slcanx-rs/examples/02_periodic.rs
use clap::Parser;
use slcanx::{Slcanx, CanFrame};
use std::time::{Duration, Instant};
use anyhow::Result;
use std::thread;

#[derive(Parser)]
struct Args {
    #[arg(short, long, default_value = "COM3")]
    port: String,
    #[arg(short, long, default_value_t = 500000)]
    bitrate: u32,
    #[arg(short, long, default_value_t = 0)]
    channel: u8,
}

fn main() -> Result<()> {
    env_logger::init();
    let args = Args::parse();

    let slcan = Slcanx::new(&args.port, 115200, Duration::from_micros(125))?;
    let channel = slcan.get_channel(args.channel);

    channel.close()?;
    channel.set_bitrate(args.bitrate)?;
    channel.open()?;

    let tx_channel = channel.clone();
    
    // Periodic sender thread
    thread::spawn(move || {
        let mut cnt = 0u8;
        loop {
            let frame = CanFrame::new_std(0x100, &[0, 1, 2, cnt]);
            if let Err(e) = tx_channel.send(frame) {
                eprintln!("Send error: {}", e);
                break;
            }
            cnt = cnt.wrapping_add(1);
            thread::sleep(Duration::from_secs(1));
        }
    });

    println!("Listening... (Ctrl+C to exit)");
    loop {
        if let Some((ch, msg)) = slcan.recv_timeout(Duration::from_millis(100))? {
            if ch == args.channel {
                println!("Received: {:?}", msg);
            }
        }
    }
}
